File: Solutions/94. Binary Tree Inorder Traversal/94. Binary Tree Inorder Traversal.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
 //recursive
class Solution {
    public List<Integer> inorderTraversal(TreeNode root) {
        var res=new ArrayList<Integer>();
        inorder(root,res);
        return res;
    }
    public void inorder(TreeNode root,List<Integer> res){
        if(root!=null){
            inorder(root.left,res);
            res.add(root.val);
            inorder(root.right,res);
        }
    }
}


// iteration stack

class Solution {
    public List<Integer> inorderTraversal(TreeNode root) {
        var stk=new Stack<TreeNode>();
        var res=new ArrayList<Integer>();
        while(!stk.empty()||root!=null){
            while(root!=null){
                stk.push(root);
                root=root.left;
            }
            root=stk.pop();
            res.add(root.val);
            root=root.right;
        }
        return res;
    }
}

File: Solutions/94. Binary Tree Inorder Traversal/94. Binary Tree Inorder Traversal.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right

//recursive
class Solution:
    def inorderTraversal(self, root: TreeNode) -> List[int]:
        def inorder(res:List[int],root:TreeNode):
            if root:
                inorder(res,root.left)
                res.append(root.val)
                inorder(res,root.right)
        res=[]
        inorder(res,root)
        return res